### ISSUE TITLE
feedback: Big-O Notation, suggestion for explanation of O(log(n))

### ISSUE THREAD

**Issue by luddd3** [robconery/feedback#5]
This is highly subjective, and please feel free to just close it if you disagree.

Even though I like your cat example I think it isn't very clear why the divide of 2 in the binary search leads to the worst case scenario O(log(n))

I googled and found this quora post which explains it further:
https://www.quora.com/How-would-you-explain-O-log-n-in-algorithms-to-1st-year-undergrad-student

I have an engineering background and math has never been my strongest subject. But reading that post I could understand how we went from "divide by 2" to O(log(n)). Sometimes the abstractions get in the way rather than help.

With that said. I think you are doing a great work with this book. Keep it up! (I also bought the red4 elixir thingy and it looks great too)

**Comment by robconery** [comment]
Yeah - a hard one to explain. So let me understand what the problem is a little more clearly and I'll see what I can come up with (btw I've read that Quora post 3 times).

The _worst_ case is O(N^2) because the complexity goes up exponentially for every _n_ operation. O(log n) is actually what you want, because execution time flattens out inversely to o(n^2).

If you close your eyes and see a graph of n^2, it curves upward, like the arms of a cactus. The exact opposite is the case with log N, it looks like an elongated lowercase "r".

It's the logarithmic splitting of the operations that achieves this (logarithm is the exact opposite of exponent). Do you think it would help if I dove into this a bit more?

**Comment by luddd3** [comment]
tl;dr: log is fine, binary search leading to log I didn't understand before seeing the math at Quora.

My main problem was that I had a hard time grasping why splitting the group into smaller groups implies O(log n). I had to see the math to understand this.

> n/2^x = 1

To translate this to cat-language =) At worst we don't find the damn cat until we have split the groups until there is only 1 left. The number of times we have to split the cat group in order to have just one cat left is x. This gives us the following worst case formula for the binary cat search:

number of cats (n) / 2^(x times) = 1 cat

The solution to this is.... x = O(log n) (as per the Quora post)

I don't think you have to elaborate more about the logarithmic function. In my case I just didn't grasp the above.

As I am writing this I went back to the cat graph that you have with the O(N^2) and O(N) lines and maybe you can include O(log n) there as well. 

I also noticed that in the graph you use **time** for the y-axis and **N** for the x-axis. In the text below it says:

> We don't want time to go up exponentially, we want our operations N to go up exponentially over linear time. That's O(log N).

Later in the section "A Quick Binary Search" you write:

> Anyway - 10 cats, 4 operations. We can keep going like this and you'll find that N (our cats) will grow exponentially compared to the operations we need to perform.

I think there might be some confusion about number of operations, execution time and size of data set.

**Comment by robconery** [comment]
Perfect explanation - I'll update! Thanks :)

**Comment by robconery** [comment]
OK - just redid a bunch of this chapter - thank you!
